### ISSUE TITLE
Load from checkpoint not working with fp16?

### ISSUE THREAD

**Issue by IsakWesterlundBitville** [NVIDIA/tacotron2#62]
Hello, I can't get loading from a checkpoint to work. 

I used this command and parameters:
```
python train.py --output_directory=outdir --log_direc
tory=logdir --checkpoint_path='outdir/checkpoint_42000'

```

```
dynamic_loss_scaling=False,
fp16_run=True,
use_saved_learning_rate=True,
batch_size=32,
```
After the first run the Loss and Grad Norm become very large as if it suddenly forgot the learning rate. The other hparams are set as default.
Could this be related to fp16 somehow?

Log:

```
FP16 Run: True
Dynamic Loss Scaling: False
Distributed Run: False
cuDNN Enabled: True
cuDNN Benchmark: False
/home/user/nv-wavenet/pytorch/tacotron2/layers.py:35: UserWarning: nn.init.xavier_uniform is now deprecated in favor of nn.init.xavier_uniform_.
  self.conv.weight, gain=torch.nn.init.calculate_gain(w_init_gain))
/home/user/nv-wavenet/pytorch/tacotron2/layers.py:15: UserWarning: nn.init.xavier_uniform is now deprecated in favor of nn.init.xavier_uniform_.
  gain=torch.nn.init.calculate_gain(w_init_gain))
FP16_Optimizer processing param group 0:
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([149, 512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024, 512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024, 256])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024, 512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024, 256])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([256, 80])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([256, 256])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096, 1536])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096, 1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([128, 1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([128, 512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1, 128])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([32, 2, 31])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([128, 32])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096, 768])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096, 1024])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([4096])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([80, 1536])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([80])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1, 1536])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([1])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 80, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([512])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([80, 512, 5])
FP16_Optimizer received torch.cuda.HalfTensor with torch.Size([80])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([80])
FP16_Optimizer received torch.cuda.FloatTensor with torch.Size([80])
Loading checkpoint 'outdir/checkpoint_42000'
Loaded checkpoint 'outdir/checkpoint_42000' from iteration 42000
Epoch: 107
train.py:220: UserWarning: invalid index of a 0-dim tensor. This will be an error in PyTorch 0.5. Use tensor.item()
 to convert a 0-dim tensor to a Python number
  if hparams.distributed_run else loss.data[0]
/home/user/nv-wavenet/pytorch/tacotron2/fp16_optimizer.py:173: UserWarning: torch.nn.utils.clip_grad_
norm is now deprecated in favor of torch.nn.utils.clip_grad_norm_.
  return torch.nn.utils.clip_grad_norm(fp32_params, clip)
Train loss 42001 0.404823 Grad Norm 0.820957 3.27s/it
OVERFLOW! Skipping step. Attempted loss scale: 262144.0
Train loss 42003 56.652672 Grad Norm 25.391602 2.56s/it
Train loss 42004 32.963730 Grad Norm 20.982843 2.57s/it
Train loss 42005 13.299093 Grad Norm 15.709991 2.64s/it
```

**Comment by rafaelvalle** [comment]
DynamicLossScaling should be on for FP16 runs.

**Comment by IsakWesterlundBitville** [comment]
Changing any parameters does not seem to make a difference. The problem is with fp16. 
Training a new model and resuming works with FP16 set to false, but when I train any model with fp16 turned on, loading from a checkpoint does not work with a Nvidia V100.

**Comment by rafaelvalle** [comment]
When checkpointing, did it get back to the expected range after a few iterations?

**Comment by gsoul** [comment]
I have the same issue and the answer for me is "no, it didn't get back to the expected range after a few iterations"

**Comment by IsakWesterlundBitville** [comment]
No, it did not for me either. Tried about 5 different configurations but all had the same problem on fp16.

**Comment by EthanKoo** [comment]
I have the same issue

**Comment by jerryli27** [comment]
+1 on this issue.

**Comment by rafaelvalle** [comment]
Closing due to inactivity.

**Comment by apsears** [comment]
@rafaelvalle Restarting on FP16 (w/ dynamic_loss_scale + FP16) wrecks the loss and grad norm, penalizing any restart by many epochs.  Is it possible to not apply gradients until the scale settles?

FP16 Run: True            
Dynamic Loss Scaling: True                                                                   
Distributed Run: False                                                 
cuDNN Enabled: True                                                                                    
cuDNN Benchmark: False                               
Selected optimization level O2:  FP16 training with FP32 batchnorm and FP32 master weights.
                                  
Defaults for this optimization level are:                                
enabled                : True
opt_level              : O2                                                       
cast_model_type        : torch.float16       
patch_torch_functions  : False                                                                                                             
keep_batchnorm_fp32    : True
master_weights         : True                      
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True                      
opt_level              : O2                                          
cast_model_type        : torch.float16                         
patch_torch_functions  : False                                          
keep_batchnorm_fp32    : True                            
master_weights         : True                                         
loss_scale             : dynamic                                
Loading checkpoint 'outdir/checkpoint_51000'
Loaded checkpoint 'outdir/checkpoint_51000' from iteration 51000
Epoch: 212                                                
Train loss 51001 0.304034 Grad Norm 0.553012 4.36s/it
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 32768.0                          
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 16384.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8192.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 4096.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 2048.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 1024.0
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 512.0
Train loss 51008 8.768527 Grad Norm inf 3.22s/it
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 256.0
Train loss 51010 8.387301 Grad Norm 58.226733 3.20s/it
Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 128.0
Train loss 51012 4.178194 Grad Norm 233.532783 3.24s/it
Train loss 51013 5.842522 Grad Norm 267.689607 3.11s/it
Train loss 51014 4.818676 Grad Norm 631.965071 2.99s/it
Train loss 51015 3.791118 Grad Norm 529.303206 3.07s/it
Train loss 51016 3.423880 Grad Norm 3536.004292 3.26s/it
Train loss 51017 4.093868 Grad Norm 15.713365 3.19s/it
Train loss 51018 4.056335 Grad Norm 13.474312 3.04s/it

**Comment by giacaglia** [comment]
Take a look at https://github.com/NVIDIA/apex#checkpointing
